htmlproofer/plugin.py

```diff
@@ -26,65 +35,81 @@ class HtmlProoferPlugin(BasePlugin):
 
     config_scheme = (
         ('raise_error', config_options.Type(bool, default=False)),
-        ('raise_error_excludes', config_options.Type(dict, default={}))
+        ('raise_error_excludes', config_options.Type(dict, default={})),
     )
 
-    def on_page_markdown(self, markdown: str, page: Page, config: Config, files: Files) -> None:
-        # Store files to allow inspecting Markdown files in later stages.
-        self.files = files
-
     def on_post_page(self, output_content: str, page: Page, config: Config) -> None:
         use_directory_urls = config.data["use_directory_urls"]
-        soup = BeautifulSoup(output_content, 'html.parser')
+
+        # Optimization: only parse links and headings
+        # li, sup are used for footnotes
+        strainer = SoupStrainer(('a', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'li', 'sup'))
+
+        soup = BeautifulSoup(output_content, 'lxml', parse_only=strainer)
+
+        all_element_ids = set(tag['id'] for tag in soup.select('[id]'))
         for a in soup.find_all('a', href=True):
             url = a['href']
-            clean_url, url_status = self.get_url_status(url, soup, self.files, use_directory_urls)
+
+            clean_url, url_status = self.get_url_status(url, all_element_ids, self.files, use_directory_urls)
+
             if self.bad_url(url_status) is True:
-                error = f'{clean_url}: {url_status}'
-                excludes = self.config['raise_error_excludes']
-                if (self.config['raise_error'] and
-                        (url_status not in excludes or
-                         ('*' not in excludes[url_status] and
-                          url not in excludes[url_status]))):
+                error = f'invalid url - {clean_url} [{url_status}] [{page.file.src_path}]'
+
+                is_error = self.is_error(self.config, url, url_status)
+                if self.config['raise_error'] and is_error:
                     raise PluginError(error)
-                else:
+                elif is_error:
                     print(error)
 
-    @lru_cache(maxsize=500)
-    def get_url_status(self, url: str, soup: BeautifulSoup, files: Files,
+    @staticmethod
+    @lru_cache(maxsize=1)
```

Comment:
What is the purpose of using an `@lru_cache` with a size of `1`?

Reply:
The default is 128, so this is to be explicit that only 1 would ever be used

Reply:
Ah, this makes sense now. Since there is no constructor in this setup this seems like an appropriate thing to do.

Reply:
Actually I think you're right that a constructor makes more sense. I was overlooking the fact that the same plugin instance is used for the lifetime of the build